lib/handlers/operatorHandler.ts

```
@@ -46,106 +49,121 @@ export class StagehandOperatorHandler {
 
-      const result = await this.getNextStep(currentStep);
+        const result = await this.getNextStep(currentStep);
 
-      if (result.method === "close") {
-        completed = true;
-      }
+        if (result.method === "close") {
+          completed = true;
+        }
 
-      let playwrightArguments: ObserveResult | undefined;
-      if (result.method === "act") {
-        [playwrightArguments] = await this.stagehandPage.page.observe(
-          result.parameters,
-        );
-      }
-      let extractionResult: unknown | undefined;
-      if (result.method === "extract") {
-        extractionResult = await this.stagehandPage.page.extract(
-          result.parameters,
-        );
+        let playwrightArguments: ObserveResult | undefined;
+        if (result.method === "act") {
+          [playwrightArguments] = await this.stagehandPage.page.observe(
+            result.parameters,
+          );
+        }
+        let extractionResult: unknown | undefined;
+        if (result.method === "extract") {
+          extractionResult = await this.stagehandPage.page.extract(
+            result.parameters,
+          );
+        }
+
+        const action: AgentAction = {
+          type: result.method,
+          reasoning: result.reasoning,
+          taskCompleted: result.taskComplete,
+          parameters: result.parameters,
+          playwrightArguments,
+          extractionResult,
+        };
+
+        await options.onStep?.(action);
+
+        await this.executeAction(result, playwrightArguments, extractionResult);
+
+        actions.push(action);
+        currentStep++;
       }
 
-      await this.executeAction(result, playwrightArguments, extractionResult);
+      const finalResult: AgentResult = {
+        success: true,
+        message: await this.getSummary(options.instruction),
+        actions,
+        completed:
+          actions.length > 0
+            ? (actions[actions.length - 1].taskCompleted as boolean)
+            : false,
```

This was a good catch/suggestion, but I'm p sure we should throw an error here not return false. The previous code was kinda catching this without being informative

---

just need to confirm cases where the agent sends a message without taking an action?